helper/users/lookup.go

```diff
@@ -30,3 +35,63 @@ func Current() (*user.User, error) {
 	defer lock.Unlock()
 	return user.Current()
 }
+
+// WriteFileFor is like os.WriteFile except if possible it chowns the file to
+// the specified user (possibly from Task.User) and sets the permissions to
+// 0o600.
+//
+// If chowning fails (either due to OS or Nomad being unprivileged), the file
+// will be left world readable (0o666).
+//
+// On failure a multierror with both the original and fallback errors will be
+// returned.
+func WriteFileFor(path string, contents []byte, username string) error {
+	// Don't even bother trying to chown to an empty username
+	var origErr error
+	if username != "" {
+		origErr := writeFileFor(path, contents, username)
+		if origErr == nil {
+			// Success!
+			return nil
+		}
+	}
+
+	// Fallback to world readable
+	if err := os.WriteFile(path, contents, 0o666); err != nil {
+		if origErr != nil {
+			// Return both errors
+			return &multierror.Error{
+				Errors: []error{origErr, err},
+			}
+		} else {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func writeFileFor(path string, contents []byte, username string) error {
+	user, err := Lookup(username)
+	if err != nil {
+		return err
+	}
+
+	uid, err := strconv.Atoi(user.Uid)
+	if err != nil {
+		return fmt.Errorf("error parsing uid: %w", err)
+	}
```

Comment:
Will this work on Windows? Where user.Uid is a SID string. I suspect where we convert to a UID in other places is always gated on unix build tags. And `chown` will always fail on Windows anyway. https://pkg.go.dev/os#Chown

Reply:
Yeah, I think this func will always fail on Windows. I could make os specific implementations, but instead I tried the approach of having a universal fallback in case of error. I'll add a Windows specific test to assert the expected behavior (fails internally, succeeds with fallback).

Reply:
The call to `Lookup` here could be quite expensive (network hops as Windows does domain lookups, etc.) but even worse it's a global mutex. The fallback is probably fine for beta but I do think we should optimize this out if we know Windows will always take the error path.

Reply:
Good point about the scary combo of `network lookup + global lock`. Wrote up #16010 - please feel free to edit/comment/fix. :grin:

Reply:
It nearly killed me, but I added a test for Windows to assert it hits the fallback behavior! :tada: It passes as both a normal user and a local Administrator because (as expected/documented) `chown` is Unix-only.